src/main/java/edu/texas/social_computing/hospitals/RHRP.java

```diff
@@ -0,0 +1,134 @@
+package edu.texas.social_computing.hospitals;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.io.FileNotFoundException;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+public class RHRP {
+    private RHRP() {
+    }
+
+    /**
+     * Real HRP with Couples From http://www.nrmp.org/couples-match-videos/
+     * - Treat couples as a unit
+     * - Only make a match if the hospital is on resident's pref list and resident is on hospital's
+     * - If one partner has failed to match, increment the other's pref rank
+     * - If one partner has been kicked out due to being the worst, then break the other partner's match (it it has one)
+     * - Assumes that both partners have same pref list size
+     */
+
+    public static Matching run(HospitalTable hospitalTable,
+                               ResidentTable residentTable,
+                               Queue<Resident> freeResidents,
+                               Matching existingMatching) {
+        Matching m = existingMatching;
+
+        while (!freeResidents.isEmpty()) {
+            Resident currentResident = freeResidents.poll();
+            if (currentResident.getPreferences().isEmpty()) {
+                if (!currentResident.hasPartner()) {
+                    continue;
+                }
+                if (residentTable.getResidentById(currentResident.getPartnerId()).getPreferences().isEmpty()) {
+                    continue;
+                }
+                // resident and partner have to have the same pref list size
+                assert (false);
+            }
+
+            Hospital hospital = hospitalTable.getHospitalById(currentResident.getPreferences()
+                    .get(residentTable.getResidentRankProgress(currentResident)));
+            residentTable.incrementResidentRankProgress(currentResident);
+
+            tryToMatch(hospital,currentResident,residentTable,freeResidents,m);
+
+            // if the currentResident managed to get assigned then try to assign partner if he has one
+            if (m.hasAssignment(currentResident) && currentResident.hasPartner()) {
+                tryToMatch(hospital,residentTable.getResidentById(currentResident.getPartnerId()),
```

Comment:
Is this trying to match the partner to the _same_ hospital? I don't think that's right... I think it should be trying to assign to the next hospital on the partner's list.